src/main/scala/io/iohk/ethereum/consensus/blocks/BlockGeneratorImpl.scala

```diff
@@ -0,0 +1,184 @@
+package io.iohk.ethereum.consensus.blocks
+
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.UnaryOperator
+
+import akka.util.ByteString
+import io.iohk.ethereum.consensus.ConsensusConfig
+import io.iohk.ethereum.consensus.ethash.blocks.Ommers
+import io.iohk.ethereum.consensus.validators.std.MptListValidator.intByteArraySerializable
+import io.iohk.ethereum.crypto.kec256
+import io.iohk.ethereum.db.dataSource.EphemDataSource
+import io.iohk.ethereum.db.storage.{ArchiveNodeStorage, NodeStorage}
+import io.iohk.ethereum.domain._
+import io.iohk.ethereum.ledger.Ledger.{BlockPreparationResult, BlockResult}
+import io.iohk.ethereum.ledger.{BlockPreparator, BloomFilter}
+import io.iohk.ethereum.mpt.{ByteArraySerializable, MerklePatriciaTrie}
+import io.iohk.ethereum.network.p2p.messages.PV62.BlockBody
+import io.iohk.ethereum.network.p2p.messages.PV62.BlockHeaderImplicits._
+import io.iohk.ethereum.utils.BlockchainConfig
+import io.iohk.ethereum.utils.ByteUtils.or
+
+import scala.collection.immutable
+
+/**
+ * This is a skeleton for a generic [[io.iohk.ethereum.consensus.blocks.BlockGenerator BlockGenerator]].
+ *
+ * @param blockchain
+ * @param blockchainConfig
+ * @param _blockTimestampProvider
+ */
+abstract class BlockGeneratorImpl(
```

Comment:
It seems this class is a collection of utility functions used by specific block generators. I don't like two things about it:
1. the name, it implies it's a standalone implementation, while it isn't
2. I can smell Java-style in it :smile: how about something along these lines:
```
trait BlockGeneratorUtils {
   def blockchain: Blockchain
   ...
   utilityFunctions()
}

class EthashBlockGenerator extends ... with BlockGeneratorUtils
```

Reply:
As the scaladoc clarifies, `This is a skeleton for a ... BlockGenerator`.

Reply:
Ok, lets leave it. I may be biased here as I usually prefer "composition-style inheritance" (with traits) rather than a abstract base implementation classes.

Reply:
https://en.wikipedia.org/wiki/Composition_over_inheritance
(I wouldn't call mixing traits "composition")

Reply:
I would agree with Lukasz about the name, though. `*Impl` is a widely recognised convention and may be confusing here. How about `BlockGeneratorSkeleton` or `BlockGeneratorPartialImpl`?

Reply:
My personal preference for many years has been `*Skeleton` for such partial implementations, so I am wondering why I did not follow it here in the first place. I think I've lost touch with the `*Impl` convention that you mention, to me it sounds like _implementation detail_ in general, whether partial or not. I tend to prepend `Std` for _standard_ implementations.

So, I can rename it to `BlockGeneratorSkeleton`.

Reply:
[Renamed](https://github.com/input-output-hk/mantis/pull/410/commits/f84265952159ab3aac2df5ded097234f31a4dc1a)